bin/helpers/utils.js

```diff
@@ -311,25 +325,187 @@ exports.getLocalFlag = (connectionSettings) => {
+  if (!this.isUndefined(args.local)) {
+    let local = false;
+    if (String(args.local).toLowerCase() === "true" || !this.isUndefined(args.local_mode)) {
+      local = true;
+      bsConfig.connection_settings.local_inferred = localInferred;
+    }
+    bsConfig["connection_settings"]["local"] = local;
+  } else if (!this.isUndefined(process.env.BROWSERSTACK_LOCAL)) {
     let local = false;
-    if (String(process.env.BROWSERSTACK_LOCAL).toLowerCase() === "true")
+    if (String(process.env.BROWSERSTACK_LOCAL).toLowerCase() === "true") {
       local = true;
+      bsConfig.connection_settings.local_inferred = localInferred;
+    }
     bsConfig["connection_settings"]["local"] = local;
     logger.info(
       "Reading local setting from the environment variable BROWSERSTACK_LOCAL"
     );
   }
 };
 
-exports.setLocalIdentifier = (bsConfig) => {
-  if (!this.isUndefined(process.env.BROWSERSTACK_LOCAL_IDENTIFIER)) {
+exports.setLocalIdentifier = (bsConfig, args) => {
+  if (!this.isUndefined(args.local_identifier)){
+    bsConfig["connection_settings"]["local_identifier"] = args.local_identifier;
+  } else if (!this.isUndefined(process.env.BROWSERSTACK_LOCAL_IDENTIFIER)) {
     bsConfig["connection_settings"]["local_identifier"] =
       process.env.BROWSERSTACK_LOCAL_IDENTIFIER;
     logger.info(
       "Reading local identifier from the environment variable BROWSERSTACK_LOCAL_IDENTIFIER"
     );
+  } else if (
+      bsConfig['connection_settings']['local'] &&
+      this.isUndefined(bsConfig["connection_settings"]["local_identifier"])
+    ){
+    bsConfig["connection_settings"]["local_identifier"] = this.generateLocalIdentifier(bsConfig['connection_settings']['local_mode']);
+  }
+};
+
+exports.setLocalMode = (bsConfig, args) => {
+  if(String(bsConfig["connection_settings"]["local"]).toLowerCase() === "true"){
+    let local_mode = 'on-demand';
+    if (!this.isUndefined(args.localMode) && args.localMode == 'always-on') {
+      local_mode = 'always-on';
+    } else if (
+      !this.isUndefined(bsConfig['connection_settings']['local_mode']) &&
+      bsConfig['connection_settings']['local_mode'].toLowerCase() ===
+        'always-on'
+    ) {
+      local_mode = 'always-on';
+    }
+    bsConfig['connection_settings']['local_mode'] = local_mode;
+    if (this.isUndefined(args.sync) || !args.sync ){
+      bsConfig['connection_settings']['sync_inferred'] = true;
+    }
+    args.sync = true;
+
+    let localModeInferred = !(this.searchForOption('--local-mode'));
+
+    if (localModeInferred) {
+      bsConfig.connection_settings.local_mode_inferred = local_mode;
+    }
+  }
+};
+
+exports.setupLocalTesting = (bsConfig, args) => {
+  return new Promise(async (resolve, reject) => {
+    let localIdentifierRunning = await this.checkLocalIdentifierRunning(
+      bsConfig, bsConfig['connection_settings']['local_identifier']
+    );
+    if (bsConfig['connection_settings']['local'] && !localIdentifierRunning){
```

Comment:
We should check the availabilty of `bsConfig['connection_settings']`

Reply:
Done